enhancements/kubelet/virtualization-higher-workload-density.md

```diff
@@ -0,0 +1,444 @@
+last-updated: "2024-06-12"
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - "https://issues.redhat.com/browse/CNV-28178"
+see-also:
+  - "https://github.com/kubernetes/enhancements/issues/2400"
+status: implementable
+---
+
+# Virtualization higher workload density
+
+## Summary
+
+Fit more workloads onto a given node - achieve a higher workload
+density - by overcommitting its memory resources. Due to timeline
+needs a multi-phased approach is considered.
+
+## Motivation
+
+Today, OpenShift Virtualization is reserving memory (`requests.memory`)
+according to the needs of the virtual machine and its infrastructure
+(the VM related pod). However, usually an application within the virtual
+machine does not utilize _all_ the memory _all_ the time. Instead,
+only _sometimes_ there are memory spikes within the virtual machine.
+And usually this is also true for the infrastucture part (the pod) of a VM:
+Not all the memory is used all the time. Because of this assumption, in
+the following we are not differentiating between the guest of a VM and the
+infrastructure of a VM, instead we are just speaking collectively of a VM.
+
+Now - Extrapolating this behavior from one to all virtual machines on a
+given node leads to the observation that _on average_ much of the reserved memory is not utilized.
+Moreover, the memory pages that are utilized can be classified to a frequently used (a.k.a working set) and inactive memory pages.
+In case of memory pressure the inactive memory pages can be swapped out. 
+From the cluster owner perspective, reserved but underutilized hardware resources - like memory in this case -
+are a cost factor.
+
+This proposal is about increasing the virtual machine density and thus
+memory utilization per node, in order to reduce the cost per virtual machine.
+
+> [!NOTE]
+> CPU resources are already over-committed in OpenShift today, thus the
+> limiting factor - usually - are memory resources.
+
+### User Stories
+
+* As an administrator, I want to be able to enable higher workload
+  density for OpenShift Virtualization clusters so that I can increase
+  the memory utilization.
+
+### Goals
+
+#### Functionality
+
+* Fit more virtual machines onto a node once higher workload density
+  is enabled
+* Integrate well with [KSM] and [FPR]
+* Protect the node from resource starvation upon high swapping activity.        
+
+#### Usability
+
+* Provide a boolean in the OpenShift Console for enabling higher-density.
+  Do not require any additional workload or cluster level configuration
+  besides the initial provisioning.
+* Provide a method to deploy the requried configuration for safer
+  memory overcommit, such as swap provisioning, tuning of system services, etc.
+* Provide a maintainable delivery method for new releases of the proposal.
+
+### Non-Goals
+
+* Allow swapping for VM pods only. We don't want to diverge from the upstream approach 
```

Comment:
isnt' this a goal until we get separate disk support in openshift?

Reply:
I don't remember that we discussed it as a condition to allow swap for all bustrable QoS pods. However we were under assumption that sticking to the upstream approach will outweigh the consideration of swap provision.
The thing is that we already released the solution in phase 2 where all the burstable pods are capable of swapping.